**Dados de telefonia móvel - Ano 2021**

**Análise quantitativa e qualitativa dos dados de 2021 (responsável: Graciele Rodrigues)**

In [1]:
# importação de bibliotecas

import pandas as pd

In [2]:
# transformando os dois arquivos em dataframe

df_2021_1 = pd.read_csv('./Dados/Acessos_Telefonia_Movel_202101-202106_Colunas.csv', sep = ';')

In [3]:
df_2021_2 = pd.read_csv('./Dados/Acessos_Telefonia_Movel_202107-202112_Colunas.csv', sep = ';')

**Dataframe - 1º semestre de 2021**

In [4]:
# df_2021_1 = df_2021_1.fillna(0)
#celula exclusiva para rodar testes

In [5]:
df_2021_1.head()

CNPJ  Código Nacional     Município  UF Modalidade de Cobrança  \
0  1778972000174               11   Santo André  SP               Pós-pago   
1  1778972000174               11   Santo André  SP               Pós-pago   
2  1778972000174               11      Alumínio  SP               Pós-pago   
3  1778972000174               11  Araçariguama  SP               Pós-pago   
4  1778972000174               11         Arujá  SP               Pós-pago   

  Tecnologia Tecnologia Geração           Empresa Porte da Prestadora  \
0        LTE                 4G  America Net S.a.       Pequeno Porte   
1        LTE                 4G  America Net S.a.       Pequeno Porte   
2        LTE                 4G  America Net S.a.       Pequeno Porte   
3        LTE                 4G  America Net S.a.       Pequeno Porte   
4        LTE                 4G  America Net S.a.       Pequeno Porte   

    Tipo de Pessoa Tipo de Produto  Código IBGE Município Grupo Econômico  \
0    Pessoa Física       VOZ+DADOS                3547809          OUTROS   
1  Pessoa Jurídica       VOZ+DADOS                3547809          OUTROS   
2    Pessoa Física       VOZ+DADOS                3501152          OUTROS   
3    Pessoa Física       VOZ+DADOS                3502754          OUTROS   
4    Pessoa Física       VOZ+DADOS                3503901          OUTROS   

   2021-01  2021-02  2021-03  2021-04  2021-05  2021-06  
0     21.0     22.0     35.0     45.0     53.0     68.0  
1     19.0     19.0     25.0     42.0     43.0     45.0  
2      2.0      2.0      2.0      2.0      2.0      3.0  
3      NaN      NaN      NaN      1.0      2.0      2.0  
4   2215.0   2297.0   2302.0   2305.0   2305.0   2332.0

In [6]:
# estudando o arquivo df_2022_1 (1º semestre) individualmente (para ver se bate com o 'merge')

print(df_2021_1.shape), '\n'
print(df_2021_1.isna().sum(), '\n')
print(df_2021_1.columns)

(395973, 19)
CNPJ                          0
Código Nacional               0
Município                     0
UF                            0
Modalidade de Cobrança        0
Tecnologia                    0
Tecnologia Geração            0
Empresa                       0
Porte da Prestadora           0
Tipo de Pessoa                0
Tipo de Produto               0
Código IBGE Município         0
Grupo Econômico               0
2021-01                   49273
2021-02                   42512
2021-03                   43661
2021-04                   42035
2021-05                   42131
2021-06                   42229
dtype: int64 

Index(['CNPJ', 'Código Nacional', 'Município', 'UF', 'Modalidade de Cobrança',
       'Tecnologia', 'Tecnologia Geração', 'Empresa', 'Porte da Prestadora',
       'Tipo de Pessoa', 'Tipo de Produto', 'Código IBGE Município',
       'Grupo Econômico', '2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06'],
      dtype='object')


In [7]:
df_2021_1['2021-01'].value_counts()

2021-01
1.0         53335
2.0         29113
3.0         19866
4.0         14818
5.0         11850
            ...  
130662.0        1
38694.0         1
29791.0         1
11450.0         1
16268.0         1
Name: count, Length: 10052, dtype: int64

**Dataframe - 2º semestre de 2021**

In [8]:
# df_2021_2 = df_2021_2.fillna(0) 
# celula exclusiva para rodar testes

In [9]:
df_2021_2.head()

CNPJ  Código Nacional     Município  UF Modalidade de Cobrança  \
0  1778972000174               11   Santo André  SP               Pós-pago   
1  1778972000174               11   Santo André  SP               Pós-pago   
2  1778972000174               11      Alumínio  SP               Pós-pago   
3  1778972000174               11  Araçariguama  SP               Pós-pago   
4  1778972000174               11         Arujá  SP               Pós-pago   

  Tecnologia Tecnologia Geração           Empresa Porte da Prestadora  \
0        LTE                 4G  America Net S.a.       Pequeno Porte   
1        LTE                 4G  America Net S.a.       Pequeno Porte   
2        LTE                 4G  America Net S.a.       Pequeno Porte   
3        LTE                 4G  America Net S.a.       Pequeno Porte   
4        LTE                 4G  America Net S.a.       Pequeno Porte   

    Tipo de Pessoa Tipo de Produto  Código IBGE Município Grupo Econômico  \
0    Pessoa Física       VOZ+DADOS                3547809          OUTROS   
1  Pessoa Jurídica       VOZ+DADOS                3547809          OUTROS   
2    Pessoa Física       VOZ+DADOS                3501152          OUTROS   
3    Pessoa Física       VOZ+DADOS                3502754          OUTROS   
4    Pessoa Física       VOZ+DADOS                3503901          OUTROS   

   2021-07  2021-08  2021-09  2021-10  2021-11  2021-12  
0     90.0     93.0    100.0    107.0     78.0     79.0  
1     53.0     55.0     65.0     74.0     75.0     65.0  
2      2.0      2.0      2.0      2.0      2.0      2.0  
3      2.0      2.0      2.0      2.0      2.0      2.0  
4   2384.0   2415.0   2438.0   2447.0   2467.0   2477.0

In [10]:
# estudando o arquivo df_2021_2 (2º semestre) individualmente (para ver se bate com o 'merge')

print(df_2021_2.shape), '\n'
print(df_2021_2.isna().sum(), '\n')
print(df_2021_2.columns)

(387037, 19)
CNPJ                          0
Código Nacional               0
Município                     0
UF                            0
Modalidade de Cobrança        0
Tecnologia                    0
Tecnologia Geração            0
Empresa                       0
Porte da Prestadora           0
Tipo de Pessoa                0
Tipo de Produto               0
Código IBGE Município         0
Grupo Econômico               0
2021-07                   33798
2021-08                   33172
2021-09                   34381
2021-10                   41096
2021-11                   41596
2021-12                   44311
dtype: int64 

Index(['CNPJ', 'Código Nacional', 'Município', 'UF', 'Modalidade de Cobrança',
       'Tecnologia', 'Tecnologia Geração', 'Empresa', 'Porte da Prestadora',
       'Tipo de Pessoa', 'Tipo de Produto', 'Código IBGE Município',
       'Grupo Econômico', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12'],
      dtype='object')


**Dataframe - Ano de 2021 (merge)**

In [11]:
df_2021 = pd.merge(df_2021_1, df_2021_2, how='outer')

In [12]:
df_2021.head(5)

CNPJ  Código Nacional     Município  UF Modalidade de Cobrança  \
0  1778972000174               11   Santo André  SP               Pós-pago   
1  1778972000174               11   Santo André  SP               Pós-pago   
2  1778972000174               11      Alumínio  SP               Pós-pago   
3  1778972000174               11  Araçariguama  SP               Pós-pago   
4  1778972000174               11         Arujá  SP               Pós-pago   

  Tecnologia Tecnologia Geração           Empresa Porte da Prestadora  \
0        LTE                 4G  America Net S.a.       Pequeno Porte   
1        LTE                 4G  America Net S.a.       Pequeno Porte   
2        LTE                 4G  America Net S.a.       Pequeno Porte   
3        LTE                 4G  America Net S.a.       Pequeno Porte   
4        LTE                 4G  America Net S.a.       Pequeno Porte   

    Tipo de Pessoa  ... 2021-03  2021-04 2021-05  2021-06  2021-07  2021-08  \
0    Pessoa Física  ...    35.0     45.0    53.0     68.0     90.0     93.0   
1  Pessoa Jurídica  ...    25.0     42.0    43.0     45.0     53.0     55.0   
2    Pessoa Física  ...     2.0      2.0     2.0      3.0      2.0      2.0   
3    Pessoa Física  ...     NaN      1.0     2.0      2.0      2.0      2.0   
4    Pessoa Física  ...  2302.0   2305.0  2305.0   2332.0   2384.0   2415.0   

   2021-09  2021-10  2021-11  2021-12  
0    100.0    107.0     78.0     79.0  
1     65.0     74.0     75.0     65.0  
2      2.0      2.0      2.0      2.0  
3      2.0      2.0      2.0      2.0  
4   2438.0   2447.0   2467.0   2477.0  

[5 rows x 25 columns]

In [13]:
print(df_2021.shape), '\n'
print(df_2021.isna().sum(), '\n')
print(df_2021.columns)

(416703, 25)
CNPJ                          0
Código Nacional               0
Município                     0
UF                            0
Modalidade de Cobrança        0
Tecnologia                    0
Tecnologia Geração            0
Empresa                       0
Porte da Prestadora           0
Tipo de Pessoa                0
Tipo de Produto               0
Código IBGE Município         0
Grupo Econômico               0
2021-01                   70003
2021-02                   63242
2021-03                   64391
2021-04                   62765
2021-05                   62861
2021-06                   62959
2021-07                   63464
2021-08                   62838
2021-09                   64047
2021-10                   70762
2021-11                   71262
2021-12                   73977
dtype: int64 

Index(['CNPJ', 'Código Nacional', 'Município', 'UF', 'Modalidade de Cobrança',
       'Tecnologia', 'Tecnologia Geração', 'Empresa', 'Porte da Prestadora',
       'Tipo de 

**Análise qualitativa dos dados (grafia dos valores)**

In [14]:
# vendo o tipo de dado que tenho por coluna
# os dados estão ok: os numéricos (especialmente os do número de acessos) estão no 
# formato float (e não object)

df_2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 416703 entries, 0 to 416702
Data columns (total 25 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   CNPJ                    416703 non-null  int64  
 1   Código Nacional         416703 non-null  int64  
 2   Município               416703 non-null  object 
 3   UF                      416703 non-null  object 
 4   Modalidade de Cobrança  416703 non-null  object 
 5   Tecnologia              416703 non-null  object 
 6   Tecnologia Geração      416703 non-null  object 
 7   Empresa                 416703 non-null  object 
 8   Porte da Prestadora     416703 non-null  object 
 9   Tipo de Pessoa          416703 non-null  object 
 10  Tipo de Produto         416703 non-null  object 
 11  Código IBGE Município   416703 non-null  int64  
 12  Grupo Econômico         416703 non-null  object 
 13  2021-01                 346700 non-null  float64
 14  2021-02             

In [15]:
# vendo o tipo de dado que tenho em cada coluna do dataframe (ou quantidade, ou qualidade), para
# ver se não há strings repetidas com diferentes grafias

# CNPJ (quantidade)
print(f"CNPJ - ", len(df_2021['CNPJ'].unique()),'\n')

# Codigo nacional (quantidade)
print(f"Código Nacional - ", len(df_2021['Código Nacional'].unique()), '\n')

# Município (quantidade)
print(f"Município - ", len(df_2021['Município'].unique()), '\n')

# UF (dados)
print(f"UF - ", len(df_2021['UF'].unique()), '\n')
print(f"UF - ", df_2021['UF'].unique(), '\n')

# Modalidade de Cobrança (quantidade)
print(f"Modalidade de Cobrança - ", len(df_2021['Modalidade de Cobrança'].unique()),'\n')

# Modalidade de Cobrança (dados)
print(f"Modalidade de Cobrança - ", df_2021['Modalidade de Cobrança'].unique(),'\n')

# Tecnologia (quantidade)
print(f"Tecnologia - ", len(df_2021['Tecnologia'].unique()), '\n')
# Tecnologia (dados)
print(f"Tecnologia - ", df_2021['Tecnologia'].unique(), '\n')

# Tecnologia Geração (quantidade)
print(f"Tecnologia Geração - ", len(df_2021['Tecnologia Geração'].unique()), '\n')
# Tecnologia Geração (dados)
print(f"Tecnologia Geração - ", df_2021['Tecnologia Geração'].unique(), '\n')

# Empresa (quantidade)
print(f"Empresa - ", len(df_2021['Empresa'].unique()), '\n')
# Empresa (dados)
print(f"Empresa - ", df_2021['Empresa'].unique(), '\n')

# Porte da Prestadora (quantidade)
print(f"Porte da Prestadora - ", len(df_2021['Porte da Prestadora'].unique()), '\n')
# Porte da Prestadora (dados)
print(f"Porte da Prestadora - ", df_2021['Porte da Prestadora'].unique(), '\n')

# Tipo de Pessoa (quantidade)
print(f"Tipo de Pessoa - ", len(df_2021['Tipo de Pessoa'].unique()), '\n')
# Tipo de Pessoa (dados)
print(f"Tipo de Pessoa - ", df_2021['Tipo de Pessoa'].unique(), '\n')

# Tipo de Produto (quantidade)
print(f"Tipo de Produto - ", len(df_2021['Tipo de Produto'].unique()), '\n')
# Tipo de Produto (dados)
print(f"Tipo de Produto - ", df_2021['Tipo de Produto'].unique(), '\n')

# Código IBGE Município (quantidade)
print(f"Código IBGE Município - ", len(df_2021['Código IBGE Município'].unique()), '\n')

# Grupo Econômico (quantidade)
print(f"Grupo Econômico - ", len(df_2021['Grupo Econômico'].unique()), '\n')
# Grupo Econômico (dados)
print(f"Grupo Econômico - ", df_2021['Grupo Econômico'].unique(), '\n')

CNPJ -  13 

Código Nacional -  67 

Município -  5297 

UF -  27 

UF -  ['SP' 'RJ' 'ES' 'MG' 'PR' 'SC' 'RS' 'GO' 'DF' 'TO' 'MT' 'MS' 'RO' 'BA'
 'SE' 'PE' 'AL' 'PB' 'RN' 'CE' 'PI' 'PA' 'AM' 'RR' 'MA' 'AC' 'AP'] 

Modalidade de Cobrança -  2 

Modalidade de Cobrança -  ['Pós-pago' 'Pré-pago'] 

Tecnologia -  5 

Tecnologia -  ['LTE' 'GSM' 'WCDMA' 'CDMA IS 95' 'NR'] 

Tecnologia Geração -  4 

Tecnologia Geração -  ['4G' '2G' '3G' '5G'] 

Empresa -  13 

Empresa -  ['America Net S.a.' 'CLARO' 'SURF TELECOM SA' 'Ligue Móvel S.A' 'NEXTEL'
 'LIGGA TELECOM' 'TIM' 'ALGAR (CTBC TELECOM)' 'VIVO' 'DATORA'
 'J. SAFRA TELECOMUNICACOES LTDA' 'OI' 'Next Level Telecom Ltda.'] 

Porte da Prestadora -  2 

Porte da Prestadora -  ['Pequeno Porte' 'Grande Porte'] 

Tipo de Pessoa -  2 

Tipo de Pessoa -  ['Pessoa Física' 'Pessoa Jurídica'] 

Tipo de Produto -  5 

Tipo de Produto -  ['VOZ+DADOS' 'M2M' 'DADOS' 'PONTO_DE_SERVICO' 'VOZ'] 

Código IBGE Município -  5570 

Grupo Econômico -  9 

Grupo Econôm

In [16]:
print(df_2021['Tecnologia'].value_counts())

Tecnologia
LTE           160509
WCDMA         134563
GSM           121145
NR               365
CDMA IS 95       121
Name: count, dtype: int64


**Análise:**

São 5570 códigos de municípios - IBGE

***Proposta: trabalhar com os códigos e não com os nomes dos municípios***

Empresas:
13 empresas
09 grupos econômicos

Tecnologia: são 5


LTE           160509
WCDMA         134563
GSM           121145
NR               365
CDMA IS 95       121

**Análise qualitativa dos dados (Not a Number)**

In [17]:
# conferindo os not a number (NaN) - ocorrência de NaN apenas nas colunas dos meses

df_2021.isna().sum()

CNPJ                          0
Código Nacional               0
Município                     0
UF                            0
Modalidade de Cobrança        0
Tecnologia                    0
Tecnologia Geração            0
Empresa                       0
Porte da Prestadora           0
Tipo de Pessoa                0
Tipo de Produto               0
Código IBGE Município         0
Grupo Econômico               0
2021-01                   70003
2021-02                   63242
2021-03                   64391
2021-04                   62765
2021-05                   62861
2021-06                   62959
2021-07                   63464
2021-08                   62838
2021-09                   64047
2021-10                   70762
2021-11                   71262
2021-12                   73977
dtype: int64

In [18]:
# Verificando a quantidade de dados nulos em cada mês - insumo para checar a qualidade dos dados

def qualidade_dados(df):
    
    print(df_2021['2021-01'].isna().value_counts(), '\n')

    print(df_2021['2021-02'].isna().value_counts(), '\n')

    print(df_2021['2021-03'].isna().value_counts(), '\n')

    print(df_2021['2021-04'].isna().value_counts(), '\n')

    print(df_2021['2021-05'].isna().value_counts(), '\n')

    print(df_2021['2021-06'].isna().value_counts(), '\n')

    print(df_2021['2021-07'].isna().value_counts(), '\n')

    print(df_2021['2021-08'].isna().value_counts(), '\n')

    print(df_2021['2021-09'].isna().value_counts(), '\n')

    print(df_2021['2021-10'].isna().value_counts(), '\n')

    print(df_2021['2021-11'].isna().value_counts(), '\n')

    print(df_2021['2021-12'].isna().value_counts(), '\n')

#print("\033[1m 2a Etapa: Combinação dos arquivos:\033[0m \n ")

In [19]:
print("True ---> dados nulos")
print("False ---> dados numéricos \n")
qualidade_dados(df_2021)

True ---> dados nulos
False ---> dados numéricos 

2021-01
False    346700
True      70003
Name: count, dtype: int64 

2021-02
False    353461
True      63242
Name: count, dtype: int64 

2021-03
False    352312
True      64391
Name: count, dtype: int64 

2021-04
False    353938
True      62765
Name: count, dtype: int64 

2021-05
False    353842
True      62861
Name: count, dtype: int64 

2021-06
False    353744
True      62959
Name: count, dtype: int64 

2021-07
False    353239
True      63464
Name: count, dtype: int64 

2021-08
False    353865
True      62838
Name: count, dtype: int64 

2021-09
False    352656
True      64047
Name: count, dtype: int64 

2021-10
False    345941
True      70762
Name: count, dtype: int64 

2021-11
False    345441
True      71262
Name: count, dtype: int64 

2021-12
False    342726
True      73977
Name: count, dtype: int64 



**Vendo os Not a Number (NaN) do 1º semestre e sua relevância face ao total de informações do dataset**

In [20]:
print("1º semestre de 2021:")

print(f"Quantidade de NaN: ", df_2021_1[['2021-01', '2021-02', '2021-03', '2021-04',
'2021-05', '2021-06']].isna().sum().sum())

# O shape do 1º semestre é
print(f"Linhas e colunas: ", df_2021_1.shape)

# numero total de entradas no meu dataframe
print(f"Total de entradas no meu df: ", df_2021_1.shape[0] *  df_2021_1.shape[1])

# vendo a percentagem no Nan na tabela do 1º semestre face ao total de informações
dado1 =  df_2021_1[['2021-01', '2021-02', '2021-03', '2021-04',
'2021-05', '2021-06']].isna().sum().sum() / ( df_2021_1.shape[0] *  df_2021_1.shape[1])
print(f"No 1º semestre de 2021, os dados NaN representam {dado1*100:.2f}% do total de entradas no dataframe")

# vendo a percentagem no Nan na tabela do 1º semestre face ao total de informações nas colunas dos meses:
dado1s =  df_2021_1[['2021-01', '2021-02', '2021-03', '2021-04',
'2021-05', '2021-06']].isna().sum().sum() / ( df_2021_1.shape[0] * 6)
print(f"No 1º semestre de 2021, os dados NaN representam {dado1s*100:.2f}% do total de entradas nas colunas de acesso (6 meses do ano)")

print("No arquivo .csv original da Anatel, esses NaN correspondem a espaços não preenchidos com valor algum: \n nem \
números, nem strings, nem caracteres especiais; correspondem a uma informação zerada")

1º semestre de 2021:
Quantidade de NaN:  261841
Linhas e colunas:  (395973, 19)
Total de entradas no meu df:  7523487
No 1º semestre de 2021, os dados NaN representam 3.48% do total de entradas no dataframe
No 1º semestre de 2021, os dados NaN representam 11.02% do total de entradas nas colunas de acesso (6 meses do ano)
No arquivo .csv original da Anatel, esses NaN correspondem a espaços não preenchidos com valor algum: 
 nem números, nem strings, nem caracteres especiais; correspondem a uma informação zerada


**Vendo os Not a Number (NaN) do 2º semestre e sua relevância face ao total de informações do dataset** 

In [21]:
print("2º semestre de 2021:")

print(f"Quantidade de NaN: ", df_2021_2[['2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']].isna().sum().sum())

# O shape do 1º semestre é
print(f"Linhas e colunas: ", df_2021_2.shape)

# numero total de entradas no meu dataframe
print(f"Total de entradas no meu df: ", df_2021_2.shape[0] * df_2021_2.shape[1])

# vendo a percentagem no Nan na tabela do 2º semestre face ao total de informações
dado2 = df_2021_2[['2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']].isna().sum().sum() / (df_2021_2.shape[0] * df_2021_2.shape[1])
print(f"No 2º semestre de 2021, os dados NaN representam {dado2*100:.2f}% do total de entradas no dataframe")

# vendo a percentagem no Nan na tabela do 2º semestre face ao total de informações nas colunas dos meses:
dado2s = df_2021_2[['2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']].isna().sum().sum() / (df_2021_2.shape[0] * 6)
print(f"No 2º semestre de 2021, os dados NaN representam {dado2s*100:.2f}% do total de entradas nas colunas de acesso (6 meses do ano)")

print("No arquivo csv original da Anatel, esses NaN correspondem a espaços não preenchidos com valor algum: \n nem \
números, nem strings, nem caracteres especiais; correspondem a uma informação zerada")

2º semestre de 2021:
Quantidade de NaN:  228354
Linhas e colunas:  (387037, 19)
Total de entradas no meu df:  7353703
No 2º semestre de 2021, os dados NaN representam 3.11% do total de entradas no dataframe
No 2º semestre de 2021, os dados NaN representam 9.83% do total de entradas nas colunas de acesso (6 meses do ano)
No arquivo csv original da Anatel, esses NaN correspondem a espaços não preenchidos com valor algum: 
 nem números, nem strings, nem caracteres especiais; correspondem a uma informação zerada


**Vendo os NaN da tabela com 'merge + outer'**

In [22]:
print("Ano de 2021:")

print(f"Quantidade de NaN: ", df_2021[['2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']].isna().sum().sum())

# O shape do ano é
print(f"Linhas e colunas: ", df_2021.shape)

# numero total de entradas no meu dataframe
print(f"Total de entradas no meu df: ", df_2021.shape[0] * df_2021.shape[1])

# vendo a percentagem no Nan na tabela 2022 face ao total de informações
dado_t = df_2021[['2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']].isna().sum().sum() / (df_2021.shape[0] * df_2021.shape[1])
print(f"No ano de 2021, os dados NaN representam {dado_t*100:.2f}% do total de entradas no dataframe")

# vendo a percentagem no Nan na tabela merge do ano face ao total de informações nas colunas dos meses:
dado_ts = df_2021[['2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']].isna().sum().sum() / (df_2021.shape[0] * 6)
print(f"No ano de 2021, os dados NaN representam {dado_ts*100:.2f}% do total de entradas nas colunas de acesso (6 meses do ano)")

Ano de 2021:
Quantidade de NaN:  792571
Linhas e colunas:  (416703, 25)
Total de entradas no meu df:  10417575
No ano de 2021, os dados NaN representam 7.61% do total de entradas no dataframe
No ano de 2021, os dados NaN representam 31.70% do total de entradas nas colunas de acesso (6 meses do ano)


**"Análise:**

Total de NaN na tabela consolidada: 3_270_405 

Deste total, 228_354 NaN vieram da tabela do 1º semestre, e 792_571 NaN vieram da tabela do 2º semestre.

No arquivo csv original da Anatel, esses NaN correspondem a espaços não preenchidos com valor algum: números, nem strings, nem caracteres especiais; correspondem a uma informação zerada, de modo que, em relação a eles, a equipe definiu que devem ser tratados tal como no arquivo original: zero.

Com o merge, recebemos novos NaN. Para saber como tráta-los e deixá-los mais fidedignos possível à análise, rodamos três testes:
1) um que pede ao python que ignore todos os NaN nos cálculos estatísticos; 
2) um que considera os NaN originais como 'zero', e ignora, como valor, os NaN provenientes do 'merge';
3) um que considera todos os NaN do dataframe como zero, passando a função df.fillna() indistintamente em todos os NaN.

Conforme abaixo, a média anual de chips ativos por empresa (considerando as variáveis tecnologia, porte etc.) que trouxe resultados mais consistentes comparativamente à análise da média mensal total por chips ativos no Brasil foi o que considera todos os NaN como zero; poir isso, rodamos nosso modelo passando a função df.fillna() em todos os NaN").

In [23]:
# fillna() apenas para fazer testes de qualidade 
# vou rodar o fillna(0) numa célula mais abaixo depois desses testes

# df_2021 = df_2021.fillna(0)

In [24]:
lista_colunas = ['2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12']

In [25]:
df_2021 = df_2021.fillna(0)

In [26]:
# criar coluna de soma_meses

df_2021["soma_meses"] = df_2021[lista_colunas].sum(axis=1)

# coluna criada para fazer o sorted e ver se há NaN relevantes dentre as empresas que figuram dentre
# as mais representativas do setor

In [27]:
# Criar coluna de média

df_2021['media_ano'] = round(df_2021[lista_colunas].mean(axis=1), 2)

In [28]:
df_2021.head(5)

CNPJ  Código Nacional     Município  UF Modalidade de Cobrança  \
0  1778972000174               11   Santo André  SP               Pós-pago   
1  1778972000174               11   Santo André  SP               Pós-pago   
2  1778972000174               11      Alumínio  SP               Pós-pago   
3  1778972000174               11  Araçariguama  SP               Pós-pago   
4  1778972000174               11         Arujá  SP               Pós-pago   

  Tecnologia Tecnologia Geração           Empresa Porte da Prestadora  \
0        LTE                 4G  America Net S.a.       Pequeno Porte   
1        LTE                 4G  America Net S.a.       Pequeno Porte   
2        LTE                 4G  America Net S.a.       Pequeno Porte   
3        LTE                 4G  America Net S.a.       Pequeno Porte   
4        LTE                 4G  America Net S.a.       Pequeno Porte   

    Tipo de Pessoa  ... 2021-05  2021-06 2021-07  2021-08  2021-09  2021-10  \
0    Pessoa Física  ...    53.0     68.0    90.0     93.0    100.0    107.0   
1  Pessoa Jurídica  ...    43.0     45.0    53.0     55.0     65.0     74.0   
2    Pessoa Física  ...     2.0      3.0     2.0      2.0      2.0      2.0   
3    Pessoa Física  ...     2.0      2.0     2.0      2.0      2.0      2.0   
4    Pessoa Física  ...  2305.0   2332.0  2384.0   2415.0   2438.0   2447.0   

   2021-11  2021-12  soma_meses  media_ano  
0     78.0     79.0       791.0      65.92  
1     75.0     65.0       580.0      48.33  
2      2.0      2.0        25.0       2.08  
3      2.0      2.0        17.0       1.42  
4   2467.0   2477.0     28384.0    2365.33  

[5 rows x 27 columns]

In [29]:
df_2021_sorted = df_2021.sort_values(by='soma_meses', ascending=True)

In [30]:
df_2021_sorted.head(10)

CNPJ  Código Nacional                Município  UF  \
416702   5423963000111               99  Vila Nova dos Martírios  MA   
132476   2421421000111               51            São Vendelino  RS   
34966   40432544000147               49              Celso Ramos  SC   
397077  40432544000147               44                   Guaíra  PR   
295851   5423963000111               22                    Carmo  RJ   
406386   2558157000162               18         Flórida Paulista  SP   
132538   2421421000111               51                  Segredo  RS   
406389   2558157000162               18          Inúbia Paulista  SP   
335839   5423963000111               54                    Áurea  RS   
320015   5423963000111               43               Carlópolis  PR   

       Modalidade de Cobrança Tecnologia Tecnologia Geração Empresa  \
416702               Pré-pago        GSM                 2G      OI   
132476               Pós-pago        LTE                 4G     TIM   
34966                Pós-pago        LTE                 4G   CLARO   
397077               Pós-pago      WCDMA                 3G   CLARO   
295851               Pós-pago      WCDMA                 3G      OI   
406386               Pós-pago      WCDMA                 3G    VIVO   
132538               Pós-pago      WCDMA                 3G     TIM   
406389               Pós-pago      WCDMA                 3G    VIVO   
335839               Pós-pago      WCDMA                 3G      OI   
320015               Pós-pago        GSM                 2G      OI   

       Porte da Prestadora   Tipo de Pessoa  ... 2021-05  2021-06 2021-07  \
416702        Grande Porte    Pessoa Física  ...     0.0      0.0     0.0   
132476        Grande Porte  Pessoa Jurídica  ...     0.0      0.0     0.0   
34966         Grande Porte  Pessoa Jurídica  ...     0.0      0.0     0.0   
397077        Grande Porte    Pessoa Física  ...     0.0      0.0     0.0   
295851        Grande Porte  Pessoa Jurídica  ...     1.0      0.0     0.0   
406386        Grande Porte    Pessoa Física  ...     0.0      0.0     0.0   
132538        Grande Porte  Pessoa Jurídica  ...     0.0      0.0     0.0   
406389        Grande Porte    Pessoa Física  ...     0.0      0.0     0.0   
335839        Grande Porte  Pessoa Jurídica  ...     1.0      0.0     0.0   
320015        Grande Porte  Pessoa Jurídica  ...     0.0      0.0     0.0   

        2021-08  2021-09  2021-10  2021-11  2021-12  soma_meses  media_ano  
416702      0.0      0.0      0.0      0.0      1.0         1.0       0.08  
132476      0.0      0.0      0.0      0.0      0.0         1.0       0.08  
34966       0.0      0.0      0.0      0.0      0.0         1.0       0.08  
397077      1.0      0.0      0.0      0.0      0.0         1.0       0.08  
295851      0.0      0.0      0.0      0.0      0.0         1.0       0.08  
406386      0.0      0.0      1.0      0.0      0.0         1.0       0.08  
132538      0.0      0.0      0.0      0.0      0.0         1.0       0.08  
406389      0.0      0.0      0.0      0.0      1.0         1.0       0.08  
335839      0.0      0.0      0.0      0.0      0.0         1.0       0.08  
320015      0.0      0.0      0.0      0.0      0.0         1.0       0.08  

[10 rows x 27 columns]

In [31]:
maiores_empresas = df_2021_sorted.tail(10)

In [32]:
maiores_empresas

CNPJ  Código Nacional       Município  UF  \
12291   40432544000147               21  Rio de Janeiro  RJ   
1934    40432544000147               11       São Paulo  SP   
173400   2558157000162               11       São Paulo  SP   
91899    2421421000111               11       São Paulo  SP   
173384   2558157000162               11       São Paulo  SP   
12278   40432544000147               21  Rio de Janeiro  RJ   
1932    40432544000147               11       São Paulo  SP   
173385   2558157000162               11       São Paulo  SP   
91920    2421421000111               11       São Paulo  SP   
1921    40432544000147               11       São Paulo  SP   

       Modalidade de Cobrança Tecnologia Tecnologia Geração Empresa  \
12291                Pré-pago        LTE                 4G   CLARO   
1934                 Pré-pago        LTE                 4G   CLARO   
173400               Pós-pago      WCDMA                 3G    VIVO   
91899                Pós-pago        LTE                 4G     TIM   
173384               Pós-pago        GSM                 2G    VIVO   
12278                Pós-pago        LTE                 4G   CLARO   
1932                 Pós-pago      WCDMA                 3G   CLARO   
173385               Pós-pago        LTE                 4G    VIVO   
91920                Pré-pago        LTE                 4G     TIM   
1921                 Pós-pago        LTE                 4G   CLARO   

       Porte da Prestadora   Tipo de Pessoa  ...    2021-05    2021-06  \
12291         Grande Porte    Pessoa Física  ...   971574.0   969547.0   
1934          Grande Porte    Pessoa Física  ...  1105024.0  1122010.0   
173400        Grande Porte  Pessoa Jurídica  ...  1282390.0  1311288.0   
91899         Grande Porte    Pessoa Física  ...  1498569.0  1481424.0   
173384        Grande Porte  Pessoa Jurídica  ...  1462878.0  1504280.0   
12278         Grande Porte    Pessoa Física  ...  1635076.0  1635307.0   
1932          Grande Porte  Pessoa Jurídica  ...  1807289.0  1767424.0   
173385        Grande Porte    Pessoa Física  ...  1825169.0  1852754.0   
91920         Grande Porte    Pessoa Física  ...  2271147.0  2259261.0   
1921          Grande Porte    Pessoa Física  ...  2491300.0  2497575.0   

          2021-07    2021-08    2021-09    2021-10    2021-11    2021-12  \
12291    978886.0   985436.0   987089.0  1358320.0  1399593.0  1000444.0   
1934    1148629.0  1163791.0  1138462.0  1456072.0  1506506.0  1161256.0   
173400  1341670.0  1348632.0  1390048.0  1164673.0  1490605.0  1426685.0   
91899   1520170.0  1511009.0  1513287.0  1522144.0  1513423.0  1490789.0   
173384  1471000.0  1441839.0  1417524.0  1455232.0  1891206.0  1873691.0   
12278   1622575.0  1634231.0  1598650.0  1664517.0  1693365.0  1576252.0   
1932    1857319.0  1735765.0  1610785.0   787116.0   819905.0  2689444.0   
173385  1861045.0  1881768.0  1910375.0  1954619.0  1893467.0  1891362.0   
91920   2360640.0  2372246.0  2371635.0  2387126.0  2363190.0  2328708.0   
1921    2472721.0  2530087.0  2459609.0  2583382.0  2635277.0  2453947.0   

        soma_meses   media_ano  
12291   12777598.0  1064799.83  
1934    14107598.0  1175633.17  
173400  15550501.0  1295875.08  
91899   18185875.0  1515489.58  
173384  18758696.0  1563224.67  
12278   19028771.0  1585730.92  
1932    19354087.0  1612840.58  
173385  21961329.0  1830110.75  
91920   28035689.0  2336307.42  
1921    29186478.0  2432206.50  

[10 rows x 27 columns]

In [33]:
# dentre as 500 maiores linhas com reportes de mais acessos, os NaN n]ao são tão óbvios, mas
# eles existem

maiores_empresas.isna().sum()

CNPJ                      0
Código Nacional           0
Município                 0
UF                        0
Modalidade de Cobrança    0
Tecnologia                0
Tecnologia Geração        0
Empresa                   0
Porte da Prestadora       0
Tipo de Pessoa            0
Tipo de Produto           0
Código IBGE Município     0
Grupo Econômico           0
2021-01                   0
2021-02                   0
2021-03                   0
2021-04                   0
2021-05                   0
2021-06                   0
2021-07                   0
2021-08                   0
2021-09                   0
2021-10                   0
2021-11                   0
2021-12                   0
soma_meses                0
media_ano                 0
dtype: int64

In [34]:
# vendo, dentro da lista com as maiores empresas (maior soma_meses), quais tem NaN em algum lugar
# do dataframe

maiores_empresas_com_nam = maiores_empresas[maiores_empresas.isna().any(axis=1)]

In [35]:
maiores_empresas_com_nam

# nesta tabela vemos que empresas com reportes relevantes tamém omitiram um mês ou outro

Empty DataFrame
Columns: [CNPJ, Código Nacional, Município, UF, Modalidade de Cobrança, Tecnologia, Tecnologia Geração, Empresa, Porte da Prestadora, Tipo de Pessoa, Tipo de Produto, Código IBGE Município, Grupo Econômico, 2021-01, 2021-02, 2021-03, 2021-04, 2021-05, 2021-06, 2021-07, 2021-08, 2021-09, 2021-10, 2021-11, 2021-12, soma_meses, media_ano]
Index: []

[0 rows x 27 columns]

In [36]:
maiores_empresas_com_nam.columns

Index(['CNPJ', 'Código Nacional', 'Município', 'UF', 'Modalidade de Cobrança',
       'Tecnologia', 'Tecnologia Geração', 'Empresa', 'Porte da Prestadora',
       'Tipo de Pessoa', 'Tipo de Produto', 'Código IBGE Município',
       'Grupo Econômico', '2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12', 'soma_meses', 'media_ano'],
      dtype='object')

In [37]:
maiores_empresas_com_nam[['Grupo Econômico', '2021-01', '2021-02', '2021-03', '2021-04',
       '2021-05', '2021-06', '2021-07', '2021-08', '2021-09', '2021-10',
       '2021-11', '2021-12', 'soma_meses', 'media_ano']]

Empty DataFrame
Columns: [Grupo Econômico, 2021-01, 2021-02, 2021-03, 2021-04, 2021-05, 2021-06, 2021-07, 2021-08, 2021-09, 2021-10, 2021-11, 2021-12, soma_meses, media_ano]
Index: []

**vendo o numero de chips (acesso) por mes, e comparando com a media**

#print(f"1ª hipotese: não fazendo fillna() nos NaN (deixando o python, por default, \n \
#ignorar o value ao calcular a média anual: \n")

print(f" Mês 2021-01 ---> {df_2021['2021-01'].sum():,.0f}\n \
Mês 2021-01 ---> {df_2021['2021-01'].sum():,.0f}\n \
Mês 2021-02 ---> {df_2021['2021-02'].sum():,.0f}\n \
Mês 2021-03 ---> {df_2021['2021-03'].sum():,.0f}\n \
Mês 2021-04 ---> {df_2021['2021-04'].sum():,.0f}\n \
Mês 2021-05 ---> {df_2021['2021-05'].sum():,.0f}\n \
Mês 2021-06 ---> {df_2021['2021-06'].sum():,.0f}\n \
Mês 2021-07 ---> {df_2021['2021-07'].sum():,.0f}\n \
Mês 2021-08 ---> {df_2021['2021-08'].sum():,.0f}\n \
Mês 2021-09 ---> {df_2021['2021-09'].sum():,.0f}\n \
Mês 2021-10 ---> {df_2021['2021-10'].sum():,.0f}\n \
Mês 2021-11 ---> {df_2021['2021-11'].sum():,.0f}\n \
Mês 2021-12 ---> {df_2021['2021-12'].sum():,.0f}\n \
 \n \
Média dos meses ---> {df_2021['media_ano'].sum():,.0f}")

#Resultado:

#1ª hipotese: não fazendo fillna() nos NaN (deixando o python, por default, 
#ignorar o value ao calcular a média anual: 

Mês 2021-01 ---> 236,014,786
Mês 2021-01 ---> 236,014,786
Mês 2021-02 ---> 238,475,286
Mês 2021-03 ---> 240,918,278
Mês 2021-04 ---> 242,108,866
Mês 2021-05 ---> 243,901,044
Mês 2021-06 ---> 245,432,123
Mês 2021-07 ---> 246,792,264
Mês 2021-08 ---> 248,495,419
Mês 2021-09 ---> 249,557,081
Mês 2021-10 ---> 251,364,359
Mês 2021-11 ---> 253,016,667
Mês 2021-12 ---> 254,711,117
  
Média dos meses ---> 249,718,424

**vendo o numero de chips (acesso) por mes, e comparando com a media**

#print(f"2ª hipotese: usando o fillna() nos NaN (o calculo da média considerará os zero nas somas dos \n \
#valores e o denominador 12 para o cálculo da média anual de cada linha: \n")

print(f" Mês 2021-01 ---> {df_2021['2021-01'].sum():,.0f}\n \
Mês 2021-01 ---> {df_2021['2021-01'].sum():,.0f}\n \
Mês 2021-02 ---> {df_2021['2021-02'].sum():,.0f}\n \
Mês 2021-03 ---> {df_2021['2021-03'].sum():,.0f}\n \
Mês 2021-04 ---> {df_2021['2021-04'].sum():,.0f}\n \
Mês 2021-05 ---> {df_2021['2021-05'].sum():,.0f}\n \
Mês 2021-06 ---> {df_2021['2021-06'].sum():,.0f}\n \
Mês 2021-07 ---> {df_2021['2021-07'].sum():,.0f}\n \
Mês 2021-08 ---> {df_2021['2021-08'].sum():,.0f}\n \
Mês 2021-09 ---> {df_2021['2021-09'].sum():,.0f}\n \
Mês 2021-10 ---> {df_2021['2021-10'].sum():,.0f}\n \
Mês 2021-11 ---> {df_2021['2021-11'].sum():,.0f}\n \
Mês 2021-12 ---> {df_2021['2021-12'].sum():,.0f}\n \
 \n \
Média dos meses ---> {df_2021['media_ano'].sum():,.0f}")

#Resultado:

#2ª hipotese: usando o fillna() nos NaN (o calculo da média considerará os zero nas somas dos 
#valores e o denominador 12 para o cálculo da média anual de cada linha: 

Mês 2021-01 ---> 236,014,786
Mês 2021-01 ---> 236,014,786
Mês 2021-02 ---> 238,475,286
Mês 2021-03 ---> 240,918,278
Mês 2021-04 ---> 242,108,866
Mês 2021-05 ---> 243,901,044
Mês 2021-06 ---> 245,432,123
Mês 2021-07 ---> 246,792,264
Mês 2021-08 ---> 248,495,419
Mês 2021-09 ---> 249,557,081
Mês 2021-10 ---> 251,364,359
Mês 2021-11 ---> 253,016,667
Mês 2021-12 ---> 254,711,117
  
Média dos meses ---> 245,898,912

In [38]:
# após a rodagem dos testes, vimos que a metodologia que melhor se aplica aos dados (e como
# eles foram organizados após o merge), vamos imputar zero a todos os NaN

df_2021 = df_2021.fillna(0)

In [39]:
pd.DataFrame(df_2021.groupby('UF').sum()['media_ano'].head(5))

media_ano
UF             
AC    805316.44
AL   3000557.22
AM   4014436.44
AP    782040.01
BA  14549399.56

In [40]:
df_2021.groupby('UF')['media_ano'].sum().head(5)

UF
AC      805316.44
AL     3000557.22
AM     4014436.44
AP      782040.01
BA    14549399.56
Name: media_ano, dtype: float64

**Verificar e corrigir as grafias dos nomes das empresas e grupos econômicos, e suas interações**

In [41]:
# criando um novo dataframe para trazer os dados do groupby: uf (unidade da federação), grupo
# grupo econômico e empresa, com a soma da variavel media_ano

# este dataframe é apenas um teste para poder visualizar os erros de grafias das empresas e 
# a atribuição de empresas e grupos econômicos

df_2021_grafia = round(pd.DataFrame(df_2021.groupby(['UF', 'Grupo Econômico', 'Empresa'])['media_ano'].sum()), 2)

In [42]:
# estados: para visualização na função query

df_2021['UF'].unique()

array(['SP', 'RJ', 'ES', 'MG', 'PR', 'SC', 'RS', 'GO', 'DF', 'TO', 'MT',
       'MS', 'RO', 'BA', 'SE', 'PE', 'AL', 'PB', 'RN', 'CE', 'PI', 'PA',
       'AM', 'RR', 'MA', 'AC', 'AP'], dtype=object)

In [43]:
df_2021['Grupo Econômico'].unique()

array(['OUTROS', 'TELECOM AMERICAS', 'NEXTEL', 'LIGGA TELECOM',
       'TELECOM ITALIA', 'ALGAR (CTBC TELECOM)', 'TELEFÔNICA', 'DATORA',
       'OI'], dtype=object)

In [44]:
# Visualizar (célula x) e corrigir (célula x)

#Criando uma função para corrigir o nome TELEFONICA 
def corrigir_nome(nome):
      if nome == 'TELEFÔNICA':
        return 'TELEFONICA'
      else:
        return nome

# Aplicando a função aos valores da coluna 'Grupo Econômico' no DataFrame
df_2021['Grupo Econômico'] = df_2021['Grupo Econômico'].apply(corrigir_nome)

In [45]:
df_2021['Grupo Econômico'].unique()

array(['OUTROS', 'TELECOM AMERICAS', 'NEXTEL', 'LIGGA TELECOM',
       'TELECOM ITALIA', 'ALGAR (CTBC TELECOM)', 'TELEFONICA', 'DATORA',
       'OI'], dtype=object)

**Fazer os dataframes por UF e Grupo Econômico**

In [46]:
df_2021_uf = round(pd.DataFrame(df_2021.groupby(['UF', 'Grupo Econômico'])['media_ano'].sum()), 2)
df_2021_uf

media_ano
UF Grupo Econômico            
AC DATORA               184.17
   NEXTEL                 4.94
   OI                 80695.14
   OUTROS              5885.24
   TELECOM AMERICAS  347934.80
...                        ...
TO OI                366428.85
   OUTROS              5911.48
   TELECOM AMERICAS  869696.04
   TELECOM ITALIA     78536.51
   TELEFONICA        237535.29

[192 rows x 1 columns]

In [47]:
# criar numa nova coluna chamada 'market_share'

df_2021_uf['market_share'] = round((df_2021_uf['media_ano'] / df_2021_uf.groupby('UF').sum()['media_ano']), 4)


In [48]:
df_2021_uf.head(30)

media_ano  market_share
UF Grupo Econômico                           
AC DATORA                184.17        0.0002
   NEXTEL                  4.94        0.0000
   OI                  80695.14        0.1002
   OUTROS               5885.24        0.0073
   TELECOM AMERICAS   347934.80        0.4320
   TELECOM ITALIA      26242.69        0.0326
   TELEFONICA         344369.46        0.4276
AL DATORA                711.67        0.0002
   NEXTEL                 27.71        0.0000
   OI                 817853.86        0.2726
   OUTROS               7263.74        0.0024
   TELECOM AMERICAS   760350.96        0.2534
   TELECOM ITALIA     926333.94        0.3087
   TELEFONICA         488015.34        0.1626
AM DATORA                540.42        0.0001
   NEXTEL                 39.76        0.0000
   OI                 523039.49        0.1303
   OUTROS              12818.74        0.0032
   TELECOM AMERICAS  1312559.85        0.3270
   TELECOM ITALIA     351621.18        0.0876
   TELEFONICA        1813817.00        0.4518
AP NEXTEL                  8.07        0.0000
   OI                 140272.66        0.1794
   OUTROS               7033.42        0.0090
   TELECOM AMERICAS   193831.80        0.2479
   TELECOM ITALIA     118114.70        0.1510
   TELEFONICA         322779.36        0.4127
BA DATORA               7320.93        0.0005
   NEXTEL                119.67        0.0000
   OI                3906404.95        0.2685

Segundo o Guia de Análise de Atos de Concentração Horizontal do Cade (“Guia H”), o “Índice Herfindahl-Hirschman
(HHI) pode ser utilizado para o cálculo do grau de concentração dos mercados. O HHI é calculado com base no
somatório do quadrado das participações de mercado de todas as empresas de um dado mercado. O HHI pode chegar
até 10.000 pontos, valor no qual há um monopólio, ou seja, em que uma única empresa possua 100% do mercado.”

Exemplo de cálculo para o Estado do Acre:

(0.14**2) + (8.34**2) + (0.73**2) + (0.37**2) + (40,61**2) + (12.55**2) + (0.16**2) + (37.10**2)

Calculando a contribuição de cada grupo economico para o calculo do HHI da região

para o cálculo do HHI, primeiro criamos uma coluna chamada hhi_empresa pegando o valor do market_share 
de cada empresa e elevando o número ao quadrado, e multiplicamos este valor por 10_000
O calculo do HHI leva em consideração a % de market share na base 100, ou seja: 40%, 30% etc
como definimos que para o Power BI era melhor deixar o market share em 4 casas decimais e sem a %
temos que multiplicar o resultado da potencia por 10_000, pois:
(40**2) = 1_600 ---> ((0.40*10^2)**2) = 1_600

In [49]:
df_2021_uf['hhi_empresa'] = (df_2021_uf['market_share'] ** 2) * 10_000

# ver se vamos chamar de hhi_empresa mesmo, ou se mudamos para hhi_grupo
# também pode ser mudado após a consolidação de todas as tabelas

In [50]:
df_2021_uf.head(20)

media_ano  market_share  hhi_empresa
UF Grupo Econômico                                        
AC DATORA                184.17        0.0002       0.0004
   NEXTEL                  4.94        0.0000       0.0000
   OI                  80695.14        0.1002     100.4004
   OUTROS               5885.24        0.0073       0.5329
   TELECOM AMERICAS   347934.80        0.4320    1866.2400
   TELECOM ITALIA      26242.69        0.0326      10.6276
   TELEFONICA         344369.46        0.4276    1828.4176
AL DATORA                711.67        0.0002       0.0004
   NEXTEL                 27.71        0.0000       0.0000
   OI                 817853.86        0.2726     743.1076
   OUTROS               7263.74        0.0024       0.0576
   TELECOM AMERICAS   760350.96        0.2534     642.1156
   TELECOM ITALIA     926333.94        0.3087     952.9569
   TELEFONICA         488015.34        0.1626     264.3876
AM DATORA                540.42        0.0001       0.0001
   NEXTEL                 39.76        0.0000       0.0000
   OI                 523039.49        0.1303     169.7809
   OUTROS              12818.74        0.0032       0.1024
   TELECOM AMERICAS  1312559.85        0.3270    1069.2900
   TELECOM ITALIA     351621.18        0.0876      76.7376

In [51]:
#a partir da coluna de hhi_empresa, somamos os valores, criando um novo dataframe com uma coluna contendo o 
# cálculo de HHI por UF ( que será a UF ou o código de área)

df_2021_hhi_uf = df_2021_uf.groupby('UF')['hhi_empresa'].sum()
df_2021_hhi_uf = pd.DataFrame(df_2021_hhi_uf).reset_index()
df_2021_hhi_uf.columns = ['UF','HHI']
df_2021_hhi_uf

# não esquecer de, em algum momento, identificar que esse dataframe é de 2021 (tipo um outro index
# em cima da coluna)

UF        HHI
0   AC  3806.2189
1   AL  2602.6257
2   AM  3357.1434
3   AP  2868.4206
4   BA  2594.6620
5   CE  2700.5774
6   DF  2961.5298
7   ES  5264.9641
8   GO  2908.2848
9   MA  2767.7267
10  MG  2956.4712
11  MS  3589.7721
12  MT  3623.3279
13  PA  2766.6636
14  PB  2784.5034
15  PE  2698.9951
16  PI  2764.7304
17  PR  3610.0336
18  RJ  2716.4603
19  RN  2858.5006
20  RO  3786.3539
21  RR  3869.6479
22  RS  3472.1840
23  SC  3235.9009
24  SE  4506.7311
25  SP  2678.7221
26  TO  3925.2799

In [52]:
#criar uma função para verificar a concentração de mercado a partir do HHI

def analise_hhi(valor_hhi):
    if valor_hhi <= 1500:
        return "pouco concentrado"
    if valor_hhi > 1500 and valor_hhi <= 2500:
        return "moderado"
    else:
        return "altamente concentrado"

In [53]:
#aplicar a função analise_hhi no dataframe, criando uma coluna analise_mercado_hhi
df_2021_hhi_uf['analise_mercado_hhi'] = df_2021_hhi_uf['HHI'].apply(analise_hhi)
df_2021_hhi_uf

UF        HHI    analise_mercado_hhi
0   AC  3806.2189  altamente concentrado
1   AL  2602.6257  altamente concentrado
2   AM  3357.1434  altamente concentrado
3   AP  2868.4206  altamente concentrado
4   BA  2594.6620  altamente concentrado
5   CE  2700.5774  altamente concentrado
6   DF  2961.5298  altamente concentrado
7   ES  5264.9641  altamente concentrado
8   GO  2908.2848  altamente concentrado
9   MA  2767.7267  altamente concentrado
10  MG  2956.4712  altamente concentrado
11  MS  3589.7721  altamente concentrado
12  MT  3623.3279  altamente concentrado
13  PA  2766.6636  altamente concentrado
14  PB  2784.5034  altamente concentrado
15  PE  2698.9951  altamente concentrado
16  PI  2764.7304  altamente concentrado
17  PR  3610.0336  altamente concentrado
18  RJ  2716.4603  altamente concentrado
19  RN  2858.5006  altamente concentrado
20  RO  3786.3539  altamente concentrado
21  RR  3869.6479  altamente concentrado
22  RS  3472.1840  altamente concentrado
23  SC  3235.9009  altamente concentrado
24  SE  4506.7311  altamente concentrado
25  SP  2678.7221  altamente concentrado
26  TO  3925.2799  altamente concentrado

**Fazer os dataframes por Código Nacional e Grupo Econômico**

In [54]:
df_2021_ddd = round(pd.DataFrame(df_2021.groupby(['Código Nacional', 'Grupo Econômico'])['media_ano'].sum()), 2)
df_2021_ddd

media_ano
Código Nacional Grupo Econômico              
11              DATORA              388445.17
                NEXTEL              111310.98
                OI                 4455817.11
                OUTROS              394658.23
                TELECOM AMERICAS  12083424.75
...                                       ...
99              OI                  443891.26
                OUTROS                4891.56
                TELECOM AMERICAS    793247.22
                TELECOM ITALIA      378327.02
                TELEFONICA          525949.54

[475 rows x 1 columns]

In [55]:
# criar numa nova coluna chamada 'market_share'

df_2021_ddd['market_share'] = round((df_2021_ddd['media_ano'] / df_2021_ddd.groupby('Código Nacional').sum()['media_ano']), 4)

In [56]:
df_2021_ddd.head(15)

media_ano  market_share
Código Nacional Grupo Econômico                            
11              DATORA              388445.17        0.0098
                NEXTEL              111310.98        0.0028
                OI                 4455817.11        0.1121
                OUTROS              394658.23        0.0099
                TELECOM AMERICAS  12083424.75        0.3040
                TELECOM ITALIA     7874052.31        0.1981
                TELEFONICA        14446428.96        0.3634
12              DATORA               20137.08        0.0058
                NEXTEL                5819.75        0.0017
                OI                  380267.67        0.1087
                OUTROS               24832.79        0.0071
                TELECOM AMERICAS    973371.53        0.2782
                TELECOM ITALIA      994994.96        0.2844
                TELEFONICA         1099715.69        0.3143
13              DATORA               15087.42        0.0051

In [57]:
# Segundo o Guia de Análise de Atos de Concentração Horizontal do Cade (“Guia H”), o “Índice Herfindahl-Hirschman
# (HHI) pode ser utilizado para o cálculo do grau de concentração dos mercados. O HHI é calculado com base no
# somatório do quadrado das participações de mercado de todas as empresas de um dado mercado. O HHI pode chegar
# até 10.000 pontos, valor no qual há um monopólio, ou seja, em que uma única empresa possua 100% do mercado.”

In [58]:
# Calculando a contribuição de cada grupo economico para o calculo do HHI da região

# para o cálculo do HHI, primeiro criamos uma coluna chamada hhi_empresa pegando o valor do market_share 
# de cada empresa e elevando o número ao quadrado, e multiplicamos este valor por 10_000
# O calculo do HHI leva em consideração a % de market share na base 100, ou seja: 40%, 30% etc
# como definimos que para o Power BI era melhor deixar o market share em 4 casas decimais e sem a %
# temos que multiplicar o resultado da potencia por 10_000, pois:
# (40**2) = 1_600 ---> ((0.40*10^2)**2) = 1_600

df_2021_ddd['hhi_empresa'] = (df_2021_ddd['market_share'] ** 2) * 10_000

# ver se vamos chamar de hhi_empresa mesmo, ou se mudamos para hhi_grupo
# também pode ser mudado após a consolidação de todas as tabelas

In [59]:
df_2021_ddd.head(10)

media_ano  market_share  hhi_empresa
Código Nacional Grupo Econômico                                         
11              DATORA              388445.17        0.0098       0.9604
                NEXTEL              111310.98        0.0028       0.0784
                OI                 4455817.11        0.1121     125.6641
                OUTROS              394658.23        0.0099       0.9801
                TELECOM AMERICAS  12083424.75        0.3040     924.1600
                TELECOM ITALIA     7874052.31        0.1981     392.4361
                TELEFONICA        14446428.96        0.3634    1320.5956
12              DATORA               20137.08        0.0058       0.3364
                NEXTEL                5819.75        0.0017       0.0289
                OI                  380267.67        0.1087     118.1569

In [60]:
#a partir da coluna de hhi_empresa, somamos os valores, criando um novo dataframe com uma coluna contendo o 
# cálculo de HHI por região (que será a UF ou o código de área)

df_2021_hhi_ddd = df_2021_ddd.groupby('Código Nacional')['hhi_empresa'].sum()
df_2021_hhi_ddd = pd.DataFrame(df_2021_hhi_ddd).reset_index()
df_2021_hhi_ddd.columns = ['DDD', 'HHI']
df_2021_hhi_ddd

# não esquecer de, em algum momento, identificar que esse dataframe é de 2022 (tipo um outro index
# em cima da coluna)

# fiquei na dívida se era para nomear a coluna como 'DDD' ou 'Código Nacional', conforme df original

DDD        HHI
0    11  2764.8747
1    12  2689.6572
2    13  2806.5961
3    14  3506.3840
4    15  3649.4411
..  ...        ...
62   95  3869.6479
63   96  2868.4206
64   97  3517.8545
65   98  2861.4822
66   99  2704.8238

[67 rows x 2 columns]

In [61]:
#criar uma função para verificar a concentração de mercado a partir do HHI

def analise_hhi(valor_hhi):
    if valor_hhi <= 1500:
        return "pouco concentrado"
    if valor_hhi > 1500 and valor_hhi <= 2500:
        return "moderado"
    else:
        return "altamente concentrado"

In [62]:
#aplicar a função analise_hhi no dataframe, criando uma coluna analise_mercado_hhi
df_2021_hhi_ddd['analise_mercado_hhi'] = df_2021_hhi_ddd['HHI'].apply(analise_hhi)
df_2021_hhi_ddd

DDD        HHI    analise_mercado_hhi
0    11  2764.8747  altamente concentrado
1    12  2689.6572  altamente concentrado
2    13  2806.5961  altamente concentrado
3    14  3506.3840  altamente concentrado
4    15  3649.4411  altamente concentrado
..  ...        ...                    ...
62   95  3869.6479  altamente concentrado
63   96  2868.4206  altamente concentrado
64   97  3517.8545  altamente concentrado
65   98  2861.4822  altamente concentrado
66   99  2704.8238  altamente concentrado

[67 rows x 3 columns]

In [63]:
df_2021['media_ano'].sum()

245898912.42

**Transformando os dataframes obtidos em arquivos 'csv'**

In [64]:
df_2021_uf.to_csv("cm_2021_uf.csv", sep=';', decimal = ",", encoding='utf-8')
try:
    print("Os dados foram exportados para o arquivo 'cm_2021_uf.csv' com sucesso.")
except Exception as e:
    print(f"Erro ao exportar os dados para o arquivo CSV: {e}")

Os dados foram exportados para o arquivo 'cm_2021_uf.csv' com sucesso.


In [65]:
df_2021_hhi_uf.to_csv("hhi_2021_uf.csv", sep=';', decimal = ",", encoding='utf-8')
try:
    print("Os dados foram exportados para o arquivo 'hhi_2021_uf.csv' com sucesso.")
except Exception as e:
    print(f"Erro ao exportar os dados para o arquivo CSV: {e}")

Os dados foram exportados para o arquivo 'hhi_2021_uf.csv' com sucesso.


In [66]:
df_2021_ddd.to_csv("cm_2021_ddd.csv", sep=';', decimal = ",", encoding='utf-8')
try:
    print("Os dados foram exportados para o arquivo 'hhi_2021_ddd.csv' com sucesso.")
except Exception as e:
    print(f"Erro ao exportar os dados para o arquivo CSV: {e}")

Os dados foram exportados para o arquivo 'hhi_2021_ddd.csv' com sucesso.


In [67]:
df_2021_hhi_ddd.to_csv("hhi_2021_ddd.csv", sep=';', decimal = ",", encoding='utf-8')
try:
    print("Os dados foram exportados para o arquivo 'hhi_2021_ddd.csv' com sucesso.")
except Exception as e:
    print(f"Erro ao exportar os dados para o arquivo CSV: {e}")

Os dados foram exportados para o arquivo 'hhi_2021_ddd.csv' com sucesso.


In [68]:
df_2021.to_csv("df_2021.csv", sep=';', decimal = ",", encoding='utf-8')
try:
    print("Os dados foram exportados para o arquivo 'df_2021.csv' com sucesso.")
except Exception as e:
    print(f"Erro ao exportar os dados para o arquivo CSV: {e}")

Os dados foram exportados para o arquivo 'df_2021.csv' com sucesso.


In [69]:
# ajuda do fábio para fazermos esta concatenação

# atribuindo outro nome ao df para não mexer nos dataframes já finalizados

df_hhi_teste = df_2021_uf.assign(HHI = lambda df: df['market_share'] ** 2).groupby('UF').sum().loc[:, ['HHI']]